### PR TITLE
Add sqlite server support

### PR DESCRIPTION
**Pull request by ekroon**
I tried implementing #356 while also taking into consideration the feedback on #414 

I ended up copying the Postgres implementation to the Sqlite implementation, which is not great for maintenance, but I did it for 2 reasons:
- A generic implementation didn't work out because I got stuck in trait boundary problems I couldn't really solve
- If I could solve those problems, the solution would be void when implementation start to drift.

#### How to use this?
`db_uri = sqlite:///my/path/to/db.db`  or other uri string as described [here](https://docs.rs/sqlx/latest/sqlx/sqlite/struct.SqliteConnectOptions.html).

#### Possible improvements:
- Instead of making the trait generic. The functions implementing the trait could be forwarded to reusable functions. I didn't try this, it might have the same compiler issues? It would save some duplicate code
- As an Sqlite setup is probably for a single user, adding configuration options which register a user could save setup time. Or alternatively, we could remove registration all together. Both solutions would make it easier to use Atuin in a Gitpod or Codespace setup by just SSH forwarding the local server port and syncing every command.

### CODE REVIEW

File: atuin-server/src/handlers/history.rs [hunk 1]
```diff
@@ -10,16 +10,18 @@ use tracing::{debug, error, instrument};
...
     calendar::{TimePeriod, TimePeriodInfo},
-    database::{Database, Postgres},
-    models::{NewHistory, User},
+    database::{DatabaseExtension, DatabaseWrapped},
+    models::{NewHistory, User as ModelUser},
 };
 
 use atuin_common::api::*;
 
+type User<T> = DatabaseWrapped<ModelUser, T>;
```
Comment (conradludgate) [inline atuin-server/src/handlers/history.rs:19]: What's the point of this exactly? I see no reason the user type needs to be coupled to the DB

Comment (ekroon) [reply atuin-server/src/handlers/history.rs:19]: 👋 @conradludgate I agree! I got stuck implementing `FromRequest` for `User` for both Postgres and Sqlite:

https://github.com/ellie/atuin/pull/491/files?diff=unified&w=0#diff-8da1109a66fb471557b348bd52fcbb14936eba4d31a11d3b1fa30b4218be1bf0R28-R31

I couldn't find a way to retrieve the correct type via `extensions` on the request. Maybe boxing the value could work?
There is no real data coupling btw, it is implemented as `PhantomData`.

If you have an idea / suggestion how to fix this I gladly take a stab at it.

**Comment by ekroon** [comment]
While implementing this feature I found: https://github.com/tokio-rs/axum/discussions/358 but I missed the link to [a dependency injection example](https://github.com/tokio-rs/axum/blob/main/examples/error-handling-and-dependency-injection/src/main.rs) in the Axum repo. 

I changed the implementation for this feature to follow that example. While this probably causes a bit of overhead (with the usage of `Arc`), the code is a lot easier to read.

**Comment by conradludgate** [comment]
I think it matters very little - async trait is already full of dynamic dispatch

**Comment by conradludgate** [comment]
Hey, sorry for the silence. With the announcement of https://tokio.rs/blog/2022-08-whats-new-in-axum-0-6-0-rc1 I've figured out a way to make it generic and not need `DynDatabase` by utilising the type-safe state.

Because of that, I'm gonna wait until axum 0.6 releases then this can be the PR to kick that off :)

**Comment by conradludgate** [comment]
On top of that, I've tested both the postgres and the sqlite DB paths. I'm very happy with this, thanks for putting in the effort :)

**Comment by evanpurkhiser** [comment]
Did anything end up blocking this?

I was interested in running my own personal atuin server in docker, but needing a postgres instance was a bit of a no start for me, so this looks awesome!

**Comment by conradludgate** [comment]
Thanks for the nudge! 

> Did anything end up blocking this?

axum 0.6, but we should get that merged soon #674

As for

> I've figured out a way to make it generic and not need DynDatabase by utilising the type-safe state

If you check out 674, you'll see that handlers take `State<AppState>`. This can actually be generic, so we can offer `Postgres` or `Sqlite` handlers each.

I'll try resume this soon when I have some spare time again, unless @ekroon is up to rebase this onto 674 and understands my idea here.

**Comment by ekroon** [comment]
@conradludgate  I think I understand and will give it a shot this weekend.

**Comment by cocoonkid** [comment]
Thank you all for this initiative. sqlite support is so much better for atuin :)

**Comment by evanpurkhiser** [comment]
Hate to nudge again but would still love to see this!

**Comment by ekroon** [comment]
I started some time ago on some integration tests as mentioned here: https://github.com/ellie/atuin/pull/711 

Will try to pick that up to get this moving again.

**Pull request closed without merging**